kernels/portable/cpu/util/test/broadcast_indexes_range_test.cpp

```diff
@@ -0,0 +1,192 @@
+TEST(BroadcastIndexesRangeTest, OneAndTwoDExhaustive) {
+  TensorFactory<ScalarType::Int> tf;
+  Tensor out = tf.zeros({3, 4});
+  Tensor in_0d_scalar = tf.zeros({});
+  Tensor in_1d_scalar = tf.zeros({1});
+  Tensor in_2d_scalar = tf.zeros({1, 1});
+
+  Tensor in_row = tf.zeros({4});
+  Tensor in_col = tf.zeros({3, 1});
+
+  Tensor in_not_broadcast = tf.zeros({3, 4});
+
+  auto actual = range_to_vec(BroadcastIndexesRange<6>(
+      out,
+      in_0d_scalar,
+      in_1d_scalar,
+      in_2d_scalar,
+      in_row,
+      in_col,
+      in_not_broadcast));
+  decltype(actual) expected = {
+      {0, 0, 0, 0, 0, 0, 0},
+      {1, 0, 0, 0, 1, 0, 1},
+      {2, 0, 0, 0, 2, 0, 2},
+      {3, 0, 0, 0, 3, 0, 3},
+      {4, 0, 0, 0, 0, 1, 4},
+      {5, 0, 0, 0, 1, 1, 5},
+      {6, 0, 0, 0, 2, 1, 6},
+      {7, 0, 0, 0, 3, 1, 7},
+      {8, 0, 0, 0, 0, 2, 8},
+      {9, 0, 0, 0, 1, 2, 9},
+      {10, 0, 0, 0, 2, 2, 10},
+      {11, 0, 0, 0, 3, 2, 11},
+  };
+  EXPECT_EQ(expected, actual);
+}
+
+// Here we assume that the previous tests established that padding
+// with leading 1s is working, and test:
+// [1, 1, 1] -> [C, H, W]
+// [C, H, 1] -> [C, H, W]
+// [C, 1, W] -> [C, H, W]
+// [1, H, W] -> [C, H, W]
+// [C, 1, 1] -> [C, H, W]
+// [1, H, 1] -> [C, H, W]
+// [1, 1, W] -> [C, H, W]
+// [C, H, W] -> [C, H, W]
+TEST(BroadcastIndexesRangeTest, ThreeDBroadcasting) {
+  TensorFactory<ScalarType::Int> tf;
+  Tensor out = tf.zeros({2, 3, 4});
+  std::array<Tensor, 8> input_tensors = {
+      tf.zeros({2, 3, 1}),
+      tf.zeros({2, 1, 4}),
+      tf.zeros({1, 3, 4}),
+      tf.zeros({2, 1, 1}),
+      tf.zeros({1, 3, 1}),
+      tf.zeros({1, 1, 4}),
+      tf.zeros({1, 1, 1}),
+      tf.zeros({2, 3, 4}),
+  };
+  // Writing out all the indexes would be too cumbersome, so here we
+  // take the opportunity to mutation test against delinearize_index
+  // and linearize_access_indexes.
+  int idx = 0;
+  for (const auto indexes : BroadcastIndexesRange<8>(
+           out,
+           input_tensors[0],
+           input_tensors[1],
+           input_tensors[2],
+           input_tensors[3],
+           input_tensors[4],
+           input_tensors[5],
+           input_tensors[6],
+           input_tensors[7])) {
+    const auto out_idx = indexes[0];
+    EXPECT_EQ(out_idx, idx++);
+    size_t out_indexes[executorch::runtime::kTensorDimensionLimit];
+    delinearize_index(
+        out_idx, out, out_indexes, executorch::runtime::kTensorDimensionLimit);
+    for (const auto tensor_idx : c10::irange(0, input_tensors.size())) {
+      EXPECT_EQ(
+          indexes[tensor_idx + 1],
+          linearize_access_indexes(
+              out_indexes, out.dim(), input_tensors[tensor_idx]));
+    }
+  }
+}
+
+// 4-D should generalize, but we will go ahead and test:
+// [N, 1, H, 1] -> [N, C, H, W]
+// [1, C, 1, W] -> [N, C, H, W]
+TEST(BroadcastIndexesRangeTest, FourDBroadcasting) {
+  TensorFactory<ScalarType::Int> tf;
+  Tensor out = tf.zeros({2, 3, 4, 5});
```

Comment:
What about out = {2, 3, 1, 5} and a = {1, 3, 1, 5} and b = {2, 1, 1, 5}

Mainly highlighting that it there is size 1 dim in the output, is it taken care of? From cursory looks I presume the answer is yes, but not sure

Reply:
>  it there is size 1 dim in the output, is it taken care of? From cursory looks I presume the answer is yes, but not sure

why would H == 1 be special? I'll add an exhaustive test for that for 1- and 2-D just in case in a follow-up.

Reply:
#8964